src/pudl/transform/sec10k.py

```diff
@@ -0,0 +1,471 @@
+
+    - per-filing company and filing information
+    - the history of company name changes
+
+    This information is used elsewhere for matching SEC 10-K filers to EIA utilities
+    and for identifying the companies that are mentioned in the unstructured Exhibit 21
+    attachements.
+
+    There is a 1-to-1 correspondence between the company_data and filing_values blocks,
+    and virtually all business_address blocks (99.9%), but a substantial number (~5%) of
+    mail_address blocks do not share index values with any company_data or filing_values
+    blocks, and are ultimately dropped. The former_company blocks contain a large amount
+    of duplicative data, but can be associated with a central_index_key value for later
+    association with an individual SEC 10-K filer.
+
+    Further processing that standardizes the contents of these tables is deferred to
+    separate downstream core assets.
+    """
+    raw_company_info = raw_sec10k__quarterly_company_information.rename(
+        columns={
+            "sec10k_filename": "filename_sec10k",
+        }
+    ).assign(
+        key=lambda x: x["key"].str.lstrip("]"),
+        filename_sec10k=lambda x: _simplify_filename_sec10k(x["filename_sec10k"]),
+    )
+
+    business_address = _pivot_info_block(
+        raw_company_info, block="business_address"
+    ).rename(columns={"business_phone": "phone"})
+    company_data = _pivot_info_block(raw_company_info, block="company_data").drop(
+        columns=["organization_name"]
+    )
+    filing_values = _pivot_info_block(raw_company_info, block="filing_values")
+    mail_address = _pivot_info_block(raw_company_info, block="mail_address")
+    former_company = _pivot_info_block(raw_company_info, block="former_company")
+
+    # Add prefixes where needed to ensure that column names are distinct after concatentation.
+    business_address.columns = [f"business_{col}" for col in business_address.columns]
+    mail_address.columns = [f"mail_{col}" for col in mail_address.columns]
+
+    # There should be a 1:1 relationship between the filing and company data.
+    assert (filing_values.sort_index().index == company_data.sort_index().index).all()
+
+    company_info = (
+        pd.merge(
+            left=company_data.sort_index(),
+            right=filing_values.sort_index(),
+            left_index=True,
+            right_index=True,
+            how="left",
+            validate="one_to_one",
+        )
+        .merge(  # 386 business addresses orphaned here.
+            business_address.sort_index(),
+            left_index=True,
+            right_index=True,
+            how="left",
+            validate="one_to_one",
+        )
+        .merge(  # 22,612 mailing addresses orphaned here.
+            mail_address.sort_index(),
+            left_index=True,
+            right_index=True,
+            how="left",
+            validate="one_to_one",
+        )
```

Comment:
Curious if @zschira understands why there would be extra records in the two address blocks that don't line up with the company info & filing values blocks.  All the records that end up merging are the ones where the filer count and block count are the same, and the ones that don't merge are those where the count values are different.

Reply:
It turns out there is an issue upstream with the way that the `block_count` is assigned -- the counter doesn't currently reset when you enter another block, so the two counters get out of sync if there's a filer that doesn't have all of the blocks (e.g. no mailing address). @zschira is going to fix it and regenerate the outputs upstream.